src/AbstractDifferentiation.jl

```diff
@@ -221,48 +221,27 @@ end
 end
 
 function pullback_function(ab::AbstractBackend, f, xs...)
-    return (ws) -> begin
-        return gradient(lowest(ab), (xs...,) -> begin
-            vs = f(xs...)
-            if ws isa Tuple
-                @assert length(vs) == length(ws)
-                return sum(Base.splat(_dot), zip(ws, vs))
-            else
-                return _dot(vs, ws)
-            end
-        end, xs...)
-    end
+    _, pbf = value_and_pullback_function(ab, f, xs...)
+    return pbf
 end
 function value_and_pullback_function(
     ab::AbstractBackend,
     f,
     xs...,
 )
-    return (ws) -> begin
-        local value
-        primalcalled = false
-        if ab isa AbstractFiniteDifference
-            value = primal_value(ab, nothing, f, xs)
-            primalcalled = true
-        end
-        if ws === nothing
-            vs = f(xs...)
-            if !primalcalled
-                value = primal_value(lowest(ab), vs, f, xs)
-                primalcalled = true
-            end
-            return value, nothing
-        end
-        pb = pullback_function(lowest(ab), (_xs...,) -> begin
+    value = f(xs...)
+    pbf = ws -> begin
+        return gradient(lowest(ab), (_xs...,) -> begin
             vs = f(_xs...)
-            if !primalcalled
-                value = primal_value(lowest(ab), vs, f, xs)
-                primalcalled = true
+            if ws isa Tuple
+                @assert length(vs) == length(ws)
```

Comment:
See #102, this is buggy

Reply:
I know :slightly_smiling_face:  My impression, however, was that it will take a bit more time to finish #102 and I don't think it is necessary to hold back this PR due to it. IMO there are too many open PRs right now, so it would be good to merge every PR that is done - even if it does not fix every currently existing bug (my general feeling is that PRs in this repo are not merged timely enough, as evident by missing reviews and #35 and #36 having become outdated).

Reply:
That makes sense. Just wanted to point it out in case @mohamed82008 wanted to review, since he's also working on the fix!

Reply:
I would be very much in favour of this more drastic change. I haven't been involved in the initial design of the package, so possibly there are very good reasons for the current design. But I've always found it confusing - also as a user - that the design contains parts such as "If f has a single input, pf_f can also accept a single input instead of a 1-tuple". In my experience, such conventions tend to become painful for both users and developers - and often lead to ambiguity issues in special cases.

Reply:
https://github.com/JuliaDiff/AbstractDifferentiation.jl/issues/103